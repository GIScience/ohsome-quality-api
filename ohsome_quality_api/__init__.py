__version__ = "1.0.0"
__title__ = "ohsome quality API"
__description__ = "Data quality estimations for OpenStreetMap"
__author__ = "ohsome team"
__email__ = "ohsome@heigit.org"
__homepage__ = "https://api.quality.ohsome.org/"
