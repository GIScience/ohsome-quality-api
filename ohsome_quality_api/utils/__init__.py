from .pybabel_yaml_extractor import pybabel_yaml_extractor

__all__ = ("pybabel_yaml_extractor",)
