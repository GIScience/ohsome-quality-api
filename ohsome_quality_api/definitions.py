"""Global Variables and Functions."""

import glob
import logging
from enum import Enum
from types import MappingProxyType
from typing import Iterable, Literal

import yaml

from ohsome_quality_api.indicators.models import IndicatorMetadata
from ohsome_quality_api.topics.definitions import load_topic_presets
from ohsome_quality_api.utils.helper import (
    camel_to_hyphen,
    get_module_dir,
)

ATTRIBUTION_TEXTS = MappingProxyType(
    {
        "OSM": "© OpenStreetMap contributors",
        "GHSL": "© European Union, 1995-2022, Global Human Settlement Topic Data",
        "VNL": "Earth Observation Group Nighttime Light Data",
        "EUBUCCO": "European building stock characteristics in a common and open "
        + "database",
        "Microsoft Buildings": "Microsoft Building Footprints (ODbL)",
    }
)

ATTRIBUTION_URL = (
    "https://github.com/GIScience/ohsome-quality-api/blob/main/COPYRIGHTS.md"
)


# default colors of the Sematic UI CSS Framework
# used by the ohsome dashboard
class Color(Enum):
    RED = "#DB2828"
    ORANGE = "#F2711C"
    YELLOW = "#FBBD08"
    OLIVE = "#B5CC18"
    GREEN = "#21BA45"
    TEAL = "#00B5AD"
    BLUE = "#2185D0"
    VIOLET = "#6435C9"
    PURPLE = "#A333C8"
    PINK = "#E03997"
    BROWN = "#A5673F"
    GREY = "#767676"
    BLACK = "#1B1C1D"


def load_metadata(
    module_name: Literal["indicators", "reports"],
) -> dict[str, IndicatorMetadata]:
    """Read metadata of all indicators or reports from YAML files.

    Those text files are located in the directory of each indicator/report.

    Args:
        module_name: Either indicators or reports.
    Returns:
        A Dict with the class names of the indicators/reports
        as keys and metadata as values.
    """
    assert module_name == "indicators" or module_name == "reports"
    directory = get_module_dir("ohsome_quality_api.{}".format(module_name))
    files = glob.glob(directory + "/**/metadata.yaml", recursive=True)
    raw = {}
    for file in files:
        with open(file, "r") as f:
            raw.update(yaml.safe_load(f))  # Merge dicts
    metadata = {}
    match module_name:
        case "indicators":
            for k, v in raw.items():
                metadata[k] = IndicatorMetadata(**v)
    return metadata


def get_metadata(
    module_name: Literal["indicators", "reports"], class_name: str
) -> IndicatorMetadata:
    """Get metadata of an indicator or report based on its class name.

    This is implemented outside the metadata class to be able to access metadata of all
    indicators/reports without instantiating of those.

    Args:
        module_name: Either indicators or reports.
        class_name: Class name of an indicator (camel case).
    """
    metadata = load_metadata(module_name)
    try:
        return metadata[camel_to_hyphen(class_name)]
    except KeyError:
        logging.error("Invalid class name: " + class_name)
        raise


# TODO: duplicate of func with the same name in projects/definition.py ?
def get_project_keys() -> Iterable[str]:
    return set(t.project for t in load_topic_presets().values())


def get_attribution(data_keys: list) -> str:
    """Return attribution text. Individual attributions are separated by semicolons."""
    assert set(data_keys) <= {"OSM", "GHSL", "VNL", "EUBUCCO", "Microsoft Buildings"}
    filtered = dict(filter(lambda d: d[0] in data_keys, ATTRIBUTION_TEXTS.items()))
    return "; ".join([str(v) for v in filtered.values()])
