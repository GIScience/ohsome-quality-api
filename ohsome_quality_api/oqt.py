"""Controller for computing Indicators."""

import logging
from typing import Coroutine

from geojson import Feature, FeatureCollection

from ohsome_quality_api.indicators.base import BaseIndicator as Indicator
from ohsome_quality_api.indicators.currentness.indicator import Currentness
from ohsome_quality_api.topics.models import BaseTopic as Topic
from ohsome_quality_api.topics.models import TopicData, TopicDefinition
from ohsome_quality_api.utils.helper import get_class_from_key
from ohsome_quality_api.utils.helper_asyncio import gather_with_semaphore
from ohsome_quality_api.utils.validators import validate_area


async def create_indicator(
    key: str,
    bpolys: FeatureCollection,
    topic: TopicData | TopicDefinition,
    include_figure: bool = True,
    **kwargs,
) -> list[Indicator]:
    """Create indicator(s) for features of a GeoJSON FeatureCollection.

    Indicators are computed asynchronously utilizing semaphores.
    Properties of the input GeoJSON are preserved.
    """
    tasks: list[Coroutine] = []
    for i, feature in enumerate(bpolys.features):
        if "id" not in feature.keys():
            feature["id"] = i
        # Disable size limit for the Mapping Saturation indicator
        # TODO: Remove size restriction
        if isinstance(topic, TopicDefinition) and key not in [
            "mapping-saturation",
            "currentness",
            "building-comparison",
            "road-comparison",
            "attribute-completeness",
            "land-cover-thematic-accuracy",
            "land-cover-completeness",
            "user-activity",
        ]:
            validate_area(feature)
        tasks.append(
            _create_indicator(
                key,
                feature,
                topic,
                include_figure,
                **kwargs,
            )
        )
    return await gather_with_semaphore(tasks)


async def _create_indicator(
    key: str,
    feature: Feature,
    topic: Topic,
    include_figure: bool = True,
    ohsomedb: bool = False,
    **kwargs,
) -> Indicator:
    """Create an indicator from scratch."""

    logging.info("Indicator key:  {0:4}".format(key))
    logging.info("Topic key:     {0:4}".format(topic.key))
    logging.info("Feature id:     {0:4}".format(feature.get("id", "None")))

    indicator_class = get_class_from_key(class_type="indicator", key=key)
    indicator = indicator_class(
        topic,
        feature,
        **kwargs,
    )

    logging.info("Run preprocessing")
    if isinstance(indicator, Currentness):
        await indicator.preprocess(ohsomedb=ohsomedb)
    else:
        await indicator.preprocess()

    logging.info("Run calculation")
    indicator.calculate()

    if include_figure:
        logging.info("Run figure creation")
        indicator.create_figure()
    else:
        indicator.result.figure = None

    return indicator
