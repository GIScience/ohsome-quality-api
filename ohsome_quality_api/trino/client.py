import asyncio
import random

import httpx
from pydantic import BaseModel, ConfigDict, field_validator

from ohsome_quality_api.utils.helper import snake_to_lower_camel

# class TrinoQuery:
#         self._query_id: Optional[str] = None
#         self._stats: Dict[Any, Any] = {}
#         self._info_uri: Optional[str] = None
#         self._warnings: List[Dict[Any, Any]] = []
#         self._columns: Optional[List[str]] = None
#         self._finished = False
#         self._cancelled = False
#         self._request = request
#         self._update_type = None
#         self._update_count = None
#         self._next_uri = None
#         self._query = query
#         self._result: Optional[TrinoResult] = None
#         self._legacy_primitive_types = legacy_primitive_types
#         self._row_mapper: Optional[RowMapper] = None
#         self._fetch_mode = fetch_mode


TRINO_HOST = "127.0.0.1"
TRINO_PORT = 8084
TRINO_USER = "mschaub"

URL = f"http://{TRINO_HOST}:{TRINO_PORT}/v1/statement"

HEADERS = {
    "X-Trino-User": TRINO_USER,
}

AUTH = None


class TrinoError(Exception):
    pass


class TrinoTimeoutError(Exception):
    pass


class BaseConfig(BaseModel):
    model_config = ConfigDict(
        alias_generator=snake_to_lower_camel,
        frozen=True,
    )


class QueryError(BaseConfig):
    message: str
    error_code: int
    error_name: str
    error_type: str
    failure_info: dict
    error_location: dict[str, int] | None = None


class QueryResults(BaseConfig):
    id: str
    next_uri: str | None = None
    columns: list | None = None
    data: list | None = None
    error: QueryError | None = None

    @field_validator("error")
    @classmethod
    def check_error(cls, value: QueryError) -> str:
        if value is not None:
            raise TrinoError(f"{value.error_name}: {value.message}")
        else:
            return value


async def query(sql) -> QueryResults:
    async with httpx.AsyncClient() as client:
        response = await client.post(
            url=URL,
            content=sql.encode("utf-8"),
            headers=HEADERS,
        )
    response.raise_for_status()
    results = QueryResults.model_validate(response.json())
    return results


async def raise_for_status(response):
    # TODO: If the client request returns an HTTP 502, 503, or 504, that means
    # there was an intermittent problem processing request and the client
    # should try again in 50-100 ms.

    # TODO: Additionally, if the request returns a 429 status code, the client should retry the request using the Retry-After header value provided.

    # TODO: Any HTTP status other than 502, 503, 504 or 200 means that query processing has failed.
    pass


# TODO: user of fetch() should be able to define a timeout instead wait+attempts.
# A possible solution is to start with very low wait times between attempts and
# scale exponentially for each failure
# TODO: Do we need backoffs and jitter?
# TODO: What is the optimal polling frequency?
async def fetch(
    results: QueryResults,
    wait: int = 1,
    attempts: int = 10,
    _attempt: int = 0,
    _data: tuple = tuple(),
) -> tuple:
    """Fetch all data of a query."""
    # TODO: use one client instance for the eitre fetching/polling
    # intervall
    async with httpx.AsyncClient() as client:
        response = await client.get(
            url=results.next_uri,
            headers=HEADERS,
        )
    response.raise_for_status()
    results = QueryResults.model_validate(response.json())

    if results.data is not None:
        _data = _data + tuple(results.data)

    if results.next_uri is None:
        return _data

    elif results.next_uri is not None and _attempt <= attempts:
        # incremental backoff + jitter
        await asyncio.sleep(wait + _attempt + random.random())
        return await fetch(results, _attempt=_attempt + 1, _data=_data)
    else:
        async with httpx.AsyncClient() as client:
            response = await client.delete(
                url=results.next_uri,
                headers=HEADERS,
            )
        raise TrinoTimeoutError()
