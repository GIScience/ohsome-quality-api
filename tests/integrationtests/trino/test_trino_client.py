import os

import pytest

# TODO: use VCR class from test utils
import vcr
from approvaltests import verify

from ohsome_quality_api.trino import client

WORKING_DIR = os.path.dirname(os.path.abspath(__file__))


@pytest.fixture
def sql_highways_count():
    file_path = os.path.join(WORKING_DIR, "highways-count.sql")
    with open(file_path, "r") as file:
        return file.read()


@pytest.fixture
def sql_highways():
    file_path = os.path.join(WORKING_DIR, "highways.sql")
    with open(file_path, "r") as file:
        return file.read()


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_query_version():
    sql = "SELECT version()"
    results = await client.query(sql)
    assert isinstance(results, client.QueryResults)
    assert results.next_uri is not None


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_fetch_version():
    sql = "SELECT version()"
    query = await client.query(sql)
    results = await client.fetch(query)
    assert isinstance(results, tuple)
    verify(str(results))


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_fetch_highways(sql_highways, sql_highways_count):
    query = await client.query(sql_highways)
    results_highways = await client.fetch(query)

    query = await client.query(sql_highways_count)
    results_highways_count = await client.fetch(query)

    assert len(results_highways) == results_highways_count[0][0]


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_query_missing_schema_name():
    sql = "SELECT * FROM foo"
    query = await client.query(sql)
    with pytest.raises(client.TrinoError) as error:
        await client.fetch(query)
    verify(str(error.value))


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_query_missing_catalog_name():
    sql = "SELECT * FROM bar.foo"
    query = await client.query(sql)
    with pytest.raises(client.TrinoError) as error:
        await client.fetch(query)
    verify(str(error.value))


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_query_catalog_not_found():
    sql = "SELECT * FROM foobar.bar.foo"
    query = await client.query(sql)
    with pytest.raises(client.TrinoError) as error:
        await client.fetch(query)
    verify(str(error.value))


@pytest.mark.asyncio()
@vcr.use_cassette
async def test_query_syntax_error():
    sql = "SELECT * MORF foo"
    query = await client.query(sql)
    with pytest.raises(client.TrinoError) as error:
        await client.fetch(query)
    verify(str(error.value))
