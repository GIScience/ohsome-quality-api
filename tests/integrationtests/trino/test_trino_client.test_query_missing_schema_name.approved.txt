MISSING_SCHEMA_NAME: line 1:15: Schema must be specified when session schema is not set
